lib/Sema/CSDiagnostics.cpp

```diff
@@ -984,9 +984,38 @@ bool MissingExplicitConversionFailure::diagnoseAsError() {
   return true;
 }
 
-bool MemberAccessOnOptionalBaseFailure::diagnoseAsError() {
+Type MemberAccessOnOptionalBaseFailure::getMemberBaseType() const {
+  if (getLocator()->isForKeyPathComponent()) {
+    if (auto *memberBaseTypeVar = MemberBaseType->getAs<TypeVariableType>()) {
+      return getSolution().getFixedType(memberBaseTypeVar);
+    }
+    return MemberBaseType;
+  }
+  return getType(getAnchor());
+}
+
+SourceRange MemberAccessOnOptionalBaseFailure::getMemberBaseSourceRange() const {
   auto anchor = getAnchor();
-  auto baseType = getType(anchor);
+  auto locator = getLocator();
+
+  if (auto componentPathElt =
+          locator->getLastElementAs<LocatorPathElt::KeyPathComponent>()) {
+    auto keyPathExpr = castToExpr<KeyPathExpr>(anchor);
+    if (componentPathElt->getIndex() == 0) {
+      return keyPathExpr->getParsedRoot()->getSourceRange();
```

Comment:
I believe you'll want to use `getRootType` instead of `getParsedRoot` to correctly handle nested type roots, e.g:

```swift
struct S {
  struct Nested {
    var foo = ""
  }
}

\S.Nested?.foo.count
```

Reply:
Ah we'll also need to handle the case where the explicit root type is null to handle cases like this:

```swift
func takesKP(_ kp: KeyPath<String?, Int>) {}
takesKP(\.count)
```